### DISCUSSION TITLE
How do i serialize a Vec<glam::Vec3> as Base64 encoded bytes ?

### DISCUSSION THREAD

**Discussion by coderedart** [jonasbb/serde_with#603] [Q&A] [answered]
I have a field with type `Vec<glam::Vec3>` to store a series of 3D positions. It would be kinda inefficient to store it as is in json. So, was wondering if i could store it as a base64 string just like we can use Base64 for a `Vec<u8>`.

I know i can do this with a manual trait impl. But was wondering if i could do this simply with a macro instead. 

Another question: I also have to base64 encode a `Uuid`. `Uuid` supports converting into `[u8; 16]`, so can i use a macro for this too?

**Comment by jonasbb** [comment, marked as answer]
You can store the `Vec3` as base64, if you know how to convert it to `Vec<u8>`. I am not familiar with glam. You will have to create your own type for which you implement `SerializeAs` and `DeserializeAs`. You can use the `Base64` from `serde_with` to save some work. Something like this:

```rust
struct Vec3Base64;

impl SerializeAs for Vec3Base64 {
    fn serialize_as<S>(source: &Vec<glam::Vec3>, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        // Fill the todo! with a way to convert the `Vec<glam::Vec3>` to `Vec<u8>`
        let bytes: Vec<u8> = todo!();
        Base64::serialize_as(bytes, serializer)
    }
}
```

The same applies to `Uuid` too. Since you can convert it into a `[u8; 16]` you can write a type with trait implementation like above that serializes a `Uuid` as `Base64`.

More information about implementing the traits is available in the documentation: 
https://docs.rs/serde_with/latest/serde_with/trait.SerializeAs.html
